src/libsync/vfs/cfapi/cfapiwrapper.cpp

```diff
@@ -124,21 +124,87 @@ void CALLBACK cfApiFetchDataCallback(const CF_CALLBACK_INFO *callbackInfo, const
     }
 
     qint64 offset = 0;
+    QByteArray protrudingData;
 
     QObject::connect(&socket, &QLocalSocket::readyRead, &loop, [&] {
-        auto data = socket.readAll();
+        const auto data = socket.readAll();
         if (data.isEmpty()) {
             qCWarning(lcCfApiWrapper) << "Unexpected empty data received" << requestId;
             sendTransferError();
             loop.quit();
             return;
         }
-        sendTransferInfo(data, offset);
-        offset += data.length();
+
+        // CFAPI expectes sended blocks to be of a multiple of a block size.
+        // Only the last sended block is allowed to be smaller than a multiple of a block size
+        // and therefore marks the end of the transmission.
+        static constexpr auto cfapiBlockSize = 4096;
+
+        // On slow connection we might not be able to fill one block complete.
+        // Threrefore we add the data to protrudingData and wait for the next run. If this
+        // was the last run, the protrudingData will get flushed after the hydration job emitted it's
+        // finished signal
+        if (data.size() + protrudingData.size() < cfapiBlockSize) {
```

Comment:
This all seems fairly complicated to me... I think if you'd start with `auto data = protudingData + dataReceived` (yes would need to rename the top one I think, and yes wouldn't be const for the) it'd end up with easier to read code. To be verified but mostly from that point it'd be either data is bigger than 4096 bytes or not. If yes data is assigned to protudingData and you bail out, if not you take the "data.size() modulo 4096" last bytes for the new protudingData and you send the rest. Or am I missing something?

Reply:
This return was because originally I sent the last protruding data in `OCC::VfsCfApi::hydrationRequestFinished`.

Reply:
Note that I meant the whole lambda, not just that if.